apps/labs/posts/scikit-lego-narwhals.md

```diff
@@ -0,0 +1,115 @@
+
+Enter narwhals. As of scikit-lego 0.9.0, you can also pass dataframes from Polars, Modin, cuDF, and in principle a whole host
+of other dataframes.
+But, how does it work? Why did they use Narwhals?
+Why don't they just
+convert to pandas internally on the user's behalf?
+
+We'll start by answering these questions, and will end with some hopes for the future.
+
+## How does it work?
+
+Let's take a look at [scikit-lego's `add_lags` function](https://koaning.github.io/scikit-lego/api/pandas-utils/?h=add_lags) as a tangible example that demonstrates how you might be able to leverage Narwhals in your own library. The code before version 0.9.0 did something like this:
+
+1. check if the input is NumPy or a pandas dataframe
+2. if it's NumPy, then perform array operations
+3. if it's pandas, then perform pandas dataframe operations
+
+In order to support Polars too, one possible solution could have been to add a fourth step to
+the above:
+
+4. if it's Polars, then perform Polars dataframe operations
+
+This looks simple, but it actually opens up a rabbit hole of extra steps, such as:
+
+5. if it's Modin, ...
+6. if it's cuDF, ...
+7. if it's fireducks, ...
+
+In order to address all the above, scikit-lego opted for a simpler approach: use Narwhals.
+The steps then become:
+
+1. run [`narwhals.from_native`](https://narwhals-dev.github.io/narwhals/api-reference/narwhals/#narwhals.from_native) on the input
+2. if it's NumPy, then perform array operations
+3. otherwise, express dataframe logic using the subset of the Polars API supported by Narwhals
+
+Like this, Narwhals takes care of translating its (Polars-like) syntax to each
+supported backend. Read the [Narwhals docs](https://narwhals-dev.github.io/narwhals) for more
+info on how to use it, and for a tutorial.
+
+## Why not "just convert to pandas"?
+
+There are different ways of converting non-pandas input to pandas. Rather than using Narwhals,
+couldn't scikit-lego just have done something like:
+
+- if the input is pandas, keep as-is
+- if the input is a non-pandas dataframe, then convert to pandas
+
+However, even in the best-case zero-copy-conversion scenario, this presents several issues:
+
+- once the input has been converted to pandas, there's no standardised way of converting back to
+  the original dataframe library.
+- using pandas as an engine may be significantly slower than doing everything using the original
+  dataframe
+
+Furthermore, converting to pandas may present a cost - for example, if you start with a Polars
+LazyFrame, which can run significantly faster because it delays reading data into memory and performs
+[several optimisations](https://docs.pola.rs/user-guide/lazy/optimizations/),
+then you're required to call [`.collect`](https://docs.pola.rs/py-polars/html/reference/lazyframe/api/polars.LazyFrame.collect.html) on it before converting to pandas.
```

Comment:
Is this like calling list() on an iterator?

Reply:
i've not heard that comparison before, but I think the analogy holds!